### PR TITLE
feat(perception): Use 2D bounding boxes for Occlusion Modifiers

### PR DESCRIPTION
**Pull request by schwepmo**
## Description

* With the addition of vehicle dimension into the `VehicleObject`, we can now calculate the bounding boxes (for now only 2D)
* The `SpatialObjectBoundingBox` interface has been added which now part of a `SpatialObject` and there exists two implementations:
    * `PointBoundingBox` -> for objects that shouldn't be treated as having a bounding box, their only "corner" will be the given position
    * `VehicleBoundingBox` -> for now used for vehicles, a rotated rectangle dependent on the heading of a vehicle.
* The changes allow for more realistic occlusion evaluation

Below are some examples of changed behavior using vehicles from the unit test:

**Vehicle Occlusion:**
<table>
    <tr>
        <th></th>
        <th>Old</th>
        <th>New></th>
    </tr>
    <tr>
        <td>#perceived vehicles</td>
        <td>27</td>
        <td>29</td>
    </tr>
    <tr>
        <td>image</td>
        <td></td>
        <td></td>
    </tr>
</table>
Quantitative there is no larger difference, with the bounding box based occlusion, actually a few more vehicles are perceived. What is noticeable, is that further non-occluded vehicles are now visible because the range penalty of the old modifier doesn't apply anymore

**Wall Occlusion:**
<table>
    <tr>
        <th></th>
        <th>Old</th>
        <th>New></th>
    </tr>
    <tr>
        <td>#perceived vehicles</td>
        <td>87</td>
        <td>86</td>
    </tr>
    <tr>
        <td>image</td>
        <td></td>
        <td></td>
    </tr>
</table>
here the difference is actually quite small. But one edge case has been randomly caught, where a vehicle crosses a wall. This vehicle is no longer visible, as 3 of its' corners are actually behind the wall.

## Issue(s) related to this PR

* Resolves internal issue 630
  
## Affected parts of the online documentation

* Yes we will need to adjust some documentation related to perception

## Definition of Done

**Prerequisites**

- [x] You have read CONTRIBUTING.md carefully.
- [x] You have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [x] Your GitHub user id is linked with your Eclipse Account.

**Required**

- [x] The title of this merge request follows the scheme `type(scope): description`  (in the style of [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary))
- [x] You have assigned a suitable label to this pull request (e.g., `enhancement`, or `bugfix`)
- [x] `origin/main` has been merged into your Fork.
- [x] Coding guidelines have been followed (see CONTRIBUTING.md).
- [x] All checks on GitHub pass.
- [x] All tests on [Jenkins](https://ci.eclipse.org/mosaic/job/mosaic/) pass.

**Requested** (can be enforced by maintainers)

- [x] New functionality is covered by unit tests or integration tests. Code coverage must not decrease.
- [ ] If a bug has been fixed, a new unit test has been written (beforehand) to prove misbehavior
- [ ] There are no new SpotBugs warnings.

### CODE REVIEW

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+import org.eclipse.mosaic.lib.math.MathUtils;
+import org.eclipse.mosaic.lib.math.Vector3d;
+import org.eclipse.mosaic.lib.math.VectorUtils;
+import org.eclipse.mosaic.lib.spatial.Edge;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class BoundingBoxOcclusionModifier implements PerceptionModifier {
```
Comment (kschrab) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java:31]: I'm questioning if we really want to call the occlusion implementations "Modifier", since they do not "modify" the perceived objects, but filter them. For me `BoundingBoxOcclusion` could be enough.

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    /**
+     * Constructor for {@link BoundingBoxOcclusionModifier}, validates and sets
+     * the parameters {@link #pointsPerSide} and {@link #detectionThreshold}.
+     *
+     * @param pointsPerSide      the amount of points that will be evaluated per object (corners count towards 2 edges)
+     * @param detectionThreshold how many points have to be visible in order for an object to be treated as detected
+     */
+    public BoundingBoxOcclusionModifier(int pointsPerSide, int detectionThreshold) {
+        if (pointsPerSide < 2) {
+            throw new RuntimeException("Need at least 2 points per edge, meaning every corner will be checked for occlusion.");
```
Comment (kschrab) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java:65]: `IllegalArgumentException` would suit better

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+    /**
+     * Threshold that defines how many of the points defined through {@link #pointsPerSide} need to be visible in order for a
+     * object to be treated as detected.
+     * Default: 2
+     */
+    private final int detectionThreshold;
+
+    /**
+     * Default constructor for the {@link BoundingBoxOcclusionModifier}.
```
Comment (kschrab) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java:49]: Add in the comment that it uses the 2,2 configuration per default.

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    public BoundingBoxOcclusionModifier() {
+        this.pointsPerSide = 2;
+        this.detectionThreshold = 2;
+    }
+
+    /**
+     * Constructor for {@link BoundingBoxOcclusionModifier}, validates and sets
+     * the parameters {@link #pointsPerSide} and {@link #detectionThreshold}.
+     *
+     * @param pointsPerSide      the amount of points that will be evaluated per object (corners count towards 2 edges)
```
Comment (kschrab) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java:60]: `... per object side`

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+            if (spatialObject instanceof TrafficLightObject) { // Traffic Lights are treated to not be occluded
+                newObjects.add(spatialObject);
+                continue;
+            }
+            List<Vector3d> pointsToEvaluate = createPointsToEvaluate(spatialObject);
+            final int requiredVisiblePoints = pointsToEvaluate.size() == 1 ? 1 : detectionThreshold;
+            int numberOfPointsVisible = 0;
+            for (Vector3d point : pointsToEvaluate) {
+                boolean pointOccluded = false;
+                boundingBoxLoop:
```
Comment (kschrab) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java:94]: instead of using labels, try to extract the loop into a new method, e.g. `boolean pointOccluded = isPointOccluded(...) `

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/index/objects/TrafficLightObject.java [hunk 6]
```diff
@@ -46,9 +50,17 @@ public TrafficLightObject(String id) {
     public TrafficLightObject setPosition(CartesianPoint position) {
         cartesianPosition.set(position);
         position.toVector3d(this);
+        if (boundingBox == null) {
+            boundingBox = new PointBoundingBox(this);
+        }
```
Comment (kschrab) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/index/objects/TrafficLightObject.java:55]: better move this to `getBoundingBox`?

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/index/objects/VehicleBoundingBox.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    private final Vector3d frontRightCorner;
+    private final Vector3d backRightCorner;
+    private final Vector3d backLeftCorner;
+    private final Vector3d frontLeftCorner;
+
+    private final List<Edge<Vector3d>> allEdges;
+    private final Edge<Vector3d> frontEdge;
+    private final Edge<Vector3d> rightEdge;
+    private final Edge<Vector3d> backEdge;
+    private final Edge<Vector3d> leftEdge;
```
Comment (kschrab) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/index/objects/VehicleBoundingBox.java:47]: for me it would be fine if these fields are `public final` and remove all getters, since they are not helpful in this case.

**Review by kschrab — APPROVED** [review]
some minor comments

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java (hunk 1 above)
Comment (schwepmo) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java:31]: I will also adjust this for the `SimpleOcclusionMordifier` -> `SimpleOcclusion` and `WallOcclusionModifer` -> `WallOcclusion`.

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/index/objects/SpatialObject.java [hunk 8]
```diff
@@ -25,14 +25,16 @@ public abstract class SpatialObject<T extends SpatialObject<T>> extends Vector3d
...
-    final MutableCartesianPoint cartesianPosition = new MutableCartesianPoint();
+    protected final MutableCartesianPoint cartesianPosition = new MutableCartesianPoint();
 
     public SpatialObject(String id) {
         this.id = id;
     }
 
     public abstract T setPosition(CartesianPoint position);
 
+    public abstract SpatialObjectBoundingBox getBoundingBox();
```
[the remark this answers is not in the document]
Comment (schwepmo) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/index/objects/SpatialObject.java:36]: Good point! I will add documentation.

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java (hunk 1 above)
Comment (schwepmo) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java:31]: We will need to adjust the documentation accordingly.

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java (hunk 5 above)
Comment (schwepmo) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java:94]: I extracted the method and also inverted the boolean logic to make the code more readable

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/index/objects/VehicleBoundingBox.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        double length = vehicleObject.getLength();
+        double halfWidth = vehicleObject.getWidth() / 2;
+        Vector3d pointA = new Vector3d(headingVector).rotateDeg(90d, VectorUtils.UP);
+        pointA = pointA.multiply(halfWidth).add(vehicleObject);
+        Vector3d pointD = new Vector3d(headingVector).rotateDeg(-90d, VectorUtils.UP);
+        pointD = pointD.multiply(halfWidth).add(vehicleObject);
+        Vector3d pointB = new Vector3d(headingVector).multiply(-1d);
+        pointB = pointB.multiply(length).add(pointA);
+        Vector3d pointC = new Vector3d(headingVector).multiply(-1d);
+        pointC = pointC.multiply(length).add(pointD);
```
[the remark this answers is not in the document]
Comment (schwepmo) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/index/objects/VehicleBoundingBox.java:74]: Good point, I documented the code.
I think I was scared of the order of operations going wrong, but you're right concatenation should be no issue.

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/index/objects/VehicleObject.java [hunk 10]
```diff
@@ -115,6 +119,23 @@ public double getHeight() {
         return height;
     }
 
+    public VehicleObject resetBoundingBox() {
```
[the remark this answers is not in the document]
Comment (schwepmo) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/index/objects/VehicleObject.java:122]: Yes, this is way nicer!

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java (hunk 1 above)
Comment (schwepmo) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java:31]: Also we could rethink naming of the other modifiers:
What do you think about:
* `PositionErrorModifier` -> `PositionError`
* `DistanceModifer` -> ??? any suggestions?

**Review by kschrab — APPROVED** [review]

**Comment by kschrab** [comment]
I just noticed that you have 2 new spotbugs warnings :<

**Comment by schwepmo** [comment]
> I just noticed that you have 2 new spotbugs warnings :<

Shame on me! I will try to fix it ASAP.

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/CentralPerceptionComponent.java [hunk 11]
```diff
@@ -225,63 +225,51 @@ static class MonitoringTrafficObjectIndexProvider extends TrafficObjectIndex {
...
-        @Override
-        public int getNumberOfVehicles() {
-            return super.getNumberOfVehicles();
-        }
-
         @Override
         public List<TrafficLightObject> getTrafficLightsInRange(PerceptionModel perceptionModel) {
             try (PerformanceMonitor.Measurement m = monitor.start("search-traffic-light")) {
-                m.setProperties(getNumberOfTrafficLights(), SimulationKernel.SimulationKernel.getCurrentSimulationTime()).restart();
+                m.setProperties(SimulationKernel.SimulationKernel.getCurrentSimulationTime()).restart();
```
Comment (kschrab) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/CentralPerceptionComponent.java:256]: I think you can remove this whole line completely. The purpose of the setProperties method when add/remove/update vehicles was, that we could track and evaluate the duration of the calls in relation to the current number of vehicles. Therefore we called `setProperties` and stored the number of vehicles + the current simulation time which was later used to generate charts for our paper. The `restart()` method was required to avoid that the call of `setProperties` was included in the performance measure.

tl;dr just remove this line completely.

**Review by kschrab — CHANGES_REQUESTED** [review]
You also need to create an MR on Extended side.

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/CentralPerceptionComponent.java [hunk 12]
```diff
@@ -225,63 +225,51 @@ static class MonitoringTrafficObjectIndexProvider extends TrafficObjectIndex {
 
         @Override
         public List<VehicleObject> getVehiclesInRange(PerceptionModel searchRange) {
+            int numberOfVehicles = getNumberOfVehicles();
             try (PerformanceMonitor.Measurement m = monitor.start("search-vehicle")) {
-                m.setProperties(getNumberOfVehicles(), SimulationKernel.SimulationKernel.getCurrentSimulationTime()).restart();
+                m.setProperties(numberOfVehicles, SimulationKernel.SimulationKernel.getCurrentSimulationTime()).restart();
```
[the remark this answers is not in the document]
Comment (schwepmo) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/CentralPerceptionComponent.java]: I reverted this. I tried to solve one spotbugs-warning that I couldn't explaub

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/CentralPerceptionComponent.java (hunk 11 above)
Comment (schwepmo) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/CentralPerceptionComponent.java:253]: We discussed this an can leave it for now but will rework this when we overthink the perception indexes.

**Review by kschrab — APPROVED** [review]
